dbt/adapters/snowflake/impl.py

```diff
@@ -147,23 +147,40 @@ def list_relations_without_caching(
         quote_policy = {"database": True, "schema": True, "identifier": True}
 
         columns = ["database_name", "schema_name", "name", "kind"]
-        for _database, _schema, _identifier, _type in results.select(columns):
-            try:
-                _type = self.Relation.get_relation_type(_type.lower())
-            except ValueError:
-                _type = self.Relation.External
+        if "is_dynamic" in results.column_names:
+            columns.append("is_dynamic")
+
+        for result in results.select(columns):
+            database, schema, identifier, relation_type = self._parse_list_relations_result(result)
             relations.append(
                 self.Relation.create(
-                    database=_database,
-                    schema=_schema,
-                    identifier=_identifier,
+                    database=database,
+                    schema=schema,
+                    identifier=identifier,
+                    type=relation_type,
                     quote_policy=quote_policy,
-                    type=_type,
                 )
             )
 
         return relations
 
+    def _parse_list_relations_result(self, result: agate.Row) -> Tuple[str, str, str, str]:
```

Comment:
The parsing bit got heftier. And in order to not duplicate the for loop above, the check needs to be done both outside of the for loop to identify the columns and inside the for loop to determine if we need to condition on that optional column. It made more sense to pull this out into its own method.

Reply:
I don't understand what you mean -- why do we need the check inside? The one outside makes enough sense.

Reply:
We need to check for whether `is_dynamic` is a field we got from the macro to produce `columns` on line 149. Then we need to condition on the value of that field, if it exists, to override Relation.Table with Relation.DynamicTable before appending the relation on line 155. At very least I needed to move the unpacking inside of the for loop to keep it a single for loop. At that point the iteration in the for loop was doing multiple checks and handling two different code paths, hence the move to its own method. Though now reading through it, I feel like I should have returned the Relation itself, instead of the pieces (basically move line 154 into this method).

Reply:
I updated this method to return the Relation that comes from the pieces this used to return. It required that I move the quote policy into this method as well, which didn't feel quite right. But taking a raw result from the list relations macro and returning a relation definitely felt more natural.